sys/sys/param.h

```diff
@@ -367,4 +367,16 @@ __END_DECLS
 #define __PAST_END(array, offset)	\
     (((__typeof__(*(array)) *)__builtin_no_change_bounds(array))[offset])
 
+#ifdef _KERNEL
+/*
+ * Convenience wrapper to convert an in-kernel pointer to a DDC-derived
+ * capability.  NB: For purecap kernels this is a no-op.
+ */
+#define	PTR2CAP(p)	({					\
+	KASSERT((vaddr_t)((p)) > VM_MAXUSER_ADDRESS,		\
```

Comment:
Hmm, should this be `>= KERNBASE` instead?

Reply:
I tried to do something portable.  INKERNEL doesn't exist on MIPS, and on some platforms INKERNEL is defined in <machine/stack.h>, not <machine/param.h> or <machine/vmparam.h>.  We also need to ensure that things like kseg0 addresses are allowed (see INKERNEL's definition on amd64 for example where the direct map is disjoint from VM_MIN/MAX_KERNEL, so we VM_MIN_KERNEL_ADDRESS wouldn't be right).  KERNBASE would be very wrong on amd64 but would work on most other platforms it seems.  Even this is wrong on i386 and powerpc with have disjoint user and kernel address spaces (for those the assertion is just wrong always).

Reply:
It seems that even VM_MAXUSER_ADDRESS isn't always defined based on the build failures.

Reply:
Probably because:
```
#if __has_feature(capabilities)
#include <machine/vmparam.h>
#endif
```
in `sys/sys/systm.h`?

Reply:
This is due to <sys/systm.h> #include'ing <machine/vmparam.h> in hybrid and purecap kernels I think.

Reply:
(which wants `VM_MAXUSER_ADDRESS` for `___USER_CFROMPTR` to null-derive things in the first page and beyond the end of userspace)

Reply:
Modulo tweaking includes, this is likely good enough for now. I mostly want to prevent bugs of the sort fixed in #626 from producing userspace capabilities that work by accident.

Reply:
I just added a stub PTR2CAP for the non-capability case.  Also, KERNBASE works for the architectures for which CHERI is currently relevant and is a bit of a tighter bound, so I also went with that.

Reply:
Oh, and `KERNBASE` doesn't work for MIPS either it turns out.  I added extra info to the panic and this is what I get on MIPS:

```
Trying to mount root from ufs:/dev/ufs/root []...
panic: PTR2CAP on user address: 0x9800000001687624 >= 0xffffffff80000000
```

KERNBASE there is not 0x8000000000000000 (VM_MIN_KERNEL_ADDRESS) but the base of K0SEG (which is sign extended on MIPS 64).  I will have to go back to testing against VM_MAXUSER_ADDRESS.